Mandatory 2FA for groups

ChristophWurst opened the pull request:
Ref https://github.com/orgs/nextcloud/projects/17.

- [x] Updated enforcement logic
- [x] Updated occ commands
- [x] Wrote tests
- [x] Requires https://github.com/nextcloud/vue-components/pull/31
- [x] Requires https://github.com/nextcloud/server/pull/11793
- [x] Updated settings UI

ChristophWurst committed:
Make it possible to enforce mandatory 2FA for groups

Signed-off-by: Christoph Wurst <christoph@winzerhof-wurst.at>

ChristophWurst commented on settings/Controller/TwoFactorSettingsController.php:
```diff
@@ -47,17 +48,15 @@ public function __construct(string $appName,
 
-	public function update(bool $enabled): Response {
-		$this->mandatoryTwoFactor->setEnforced($enabled);
+	public function update(bool $enforced, array $enforcedGroups = [], array $excludedGroups = []): Response {
```
I know. I deliberately chose the base response class to stay compatible if we ever make controllers strict :wink:

rullzer replied:
But if they are strict that still works

ChristophWurst replied:
```php
<?php

declare(strict_types=1);

class A {}
class B extends A {}

class C {
	public function foo(): A { return new A(); }
}

class D extends C {
	public function foo(): B { return new B(); }
}
```

```
php -f strict.php 
PHP Fatal error:  Declaration of D::foo(): B must be compatible with C::foo(): A in strict.php on line 14
```

That's the use case I was referring to: making resource controller methods strict. But now on a second thought it won't work anyway, as we can't have abstract declarations of `index` and such because we can't know/force the parameters.

I'll switch to `JSONResponse` type hints then.

ChristophWurst commented:
This works quite well, let's get that multiselect component in :)

![bildschirmfoto von 2018-10-12 11-27-43](https://user-images.githubusercontent.com/1374172/46861176-55dbc700-ce12-11e8-9eb7-ad211baa819e.png)

rullzer commented:
1. Open security settings
2. Have 2fa not enforces and have no groups configured
3. Press on the enforce 2fa for groups thingy
4. Expect to have groups listed

It just shows me The list is empty


------

1. Open security settings
2. Have 2fa not enforces and have no groups configured
3. Press on the exclude groups
4. Nothing is listed or whatever is entered doesn't do anything

This should be disabled I guess if 2FA is not enforced for anybody?

ChristophWurst commented:
> It just shows me The list is empty

Fixed.

> This should be disabled I guess if 2FA is not enforced for anybody?

Fixed. I deliberately did not disable the group inputs when 2FA is not enforced. This allows admins to properly configure enforced 2FA before it gets activated. Otherwise users might have login problems between the enforcement of 2FA and the proper configuration of the enforced groups.

rullzer reviewed (approved):
Clicked around a bit with different scenarios. Does the trick for me.
Lets get this in :rocket:

ChristophWurst commented:
>     * do not show group settings if the enforce setting is not enabled

See https://github.com/nextcloud/server/pull/11765#issuecomment-429713456. I can change this, however.

>     * somehow the placeholder text in the group fields is missing a left padding (once clicking inside the field it is correctly placed:

cc @juliushaertl I guess we have to fix this in the vue components lib then.

MorrisJobke commented:
> cc @juliushaertl I guess we have to fix this in the vue components lib then.

Okay - fine with that then 👍
